### PR TITLE
Implement the KimYi algorithm

### PR DESCRIPTION
**Pull request by ltratt**
The first 4 commits mostly just move code around semi-mechanically: the final commit is where the action lies. It is quite big, unfortunately, but it's really difficult to split it into smaller bits.

After playing with this for weeks, I think I've understood the paper well enough to derive a fairly faithful implementation (bearing in mind that the paper is somewhat incomplete, and frequently confusing). As the paper suggests, this only finds a single minimal cost solution to each parser error, so it's highly non-deterministic: sometimes the minimal cost solution it comes up with is good, and sometimes it isn't. It is, however, pretty fast, certainly compared to the Corchuelo et al. algorithm. The biggest extension relative to the paper is that they really only care about finding a repair: they don't seem to care about reporting that to the user. The repairs thus reference non-terminals in the grammar which, to most users, are completely incomprehensible. This commit uses cfgrammar's SentenceGenerator to turn a non-terminal repair into a (possible set) of terminal inserts, which is infinitely easier to understand. Thus whilst KimYi might say a repair is:

```
  InsertNonterm expr
```

this commit will say something like:

```
  InsertTerm {Var, Identifier}, InsertTerm{+, -, *}, InsertTerm {Var, Identifier}
```

There are several glaring inefficiencies in this commit: notably, we continually make new SentenceGenerators. That's because, currently, we don't really have a good way for passing state between, and within, error recovery instances. However, that's a more invasive change which can come later.

To test this on a real Java file use something like:

```
$ cargo run --release java.l java.y <java file with errors in>
```

which will print the parse tree and then repairs e.g.:

```
Error detected at line 56 col 13. Amongst the valid repairs are:
  Insert "EQ"
Error detected at line 56 col 62. Amongst the valid repairs are:
  Insert {INTEGER_LITERAL, FLOATING_POINT_LITERAL, BOOLEAN_LITERAL, CHARACTER_LITERAL, STRING_LITERAL, NULL_LITERAL, THIS, IDENTIFIER}, Keep ")", Insert "SEMICOLON"
Error detected at line 1137 col 41. Amongst the valid repairs are:
  Insert "AND", Keep "to", Insert "QUESTION"
```

The big difference relative to the Corchuelo et al. algorithm is that you can end up with "choose from a set". e.g. the second repair says "choose any of {integer, float, ..., identifier}".

A future PR will implement what I'm tentatively calling the KimYi+ algorithm, which is basically their algorithm plus various ideas I've collected to make it more useful to humans. That might slightly depend on how https://github.com/samueltardieu/pathfinding/pull/44 ends up going.

### CODE REVIEW

File: src/lib/corchuelo.rs [hunk 1]
```diff
@@ -88,7 +88,7 @@ pub(crate) fn recover<TokId: Clone + Copy + Debug + TryFrom<usize> + TryInto<usi
             },
             _ => {
                 let num_inserts = repairs.vals()
-                                         .filter(|r| if let ParseRepair::InsertTerm{..} = **r {
+                                         .filter(|r| if let ParseRepair::Insert{..} = **r {
```
Comment (ptersilie) [inline src/lib/corchuelo.rs]: Why did you rename these in the second commit, only to rename them back here?

Comment (ltratt) [reply src/lib/corchuelo.rs]: Hmm, that's a rebasing gone wrong (my fault). I can fix things if you want before you do a full review?

Comment (ptersilie) [reply src/lib/corchuelo.rs]: I'll try to finish my review first.

Comment (ptersilie) [reply src/lib/corchuelo.rs]: I'm surprised this didn't result in an error.

Comment (ltratt) [reply src/lib/corchuelo.rs]: It's not an error-error, just a mistake in the rebasing on my part I think.

Comment (ptersilie) [reply src/lib/corchuelo.rs]: Just saw, this get's refixed in the next commit which is why the tests don't catch this.

File: src/lib/kimyi.rs [hunk 2]
```diff
@@ -0,0 +1,737 @@
...
+
+pub(crate) fn recover<TokId: Clone + Copy + Debug + TryFrom<usize> + TryInto<usize> + PartialEq>
+                     (parser: &Parser<TokId>, in_la_idx: usize, in_pstack: &mut Vec<StIdx>,
+                      mut tstack: &mut Vec<Node<TokId>>)
+                  -> (usize, Vec<Vec<ParseRepair>>)
+{
+    // This function implements the algorithm from "LR error repair using the A* algorithm" by
+    // Ik-Soon Kim and Kwangkeun Yi
+    //
+    // The basic idea is to define finding repairs as finding the shortest route through a graph to
```
Comment (ptersilie) [inline src/lib/kimyi.rs]: random `finding`

Comment (ltratt) [reply src/lib/kimyi.rs]: It's sort-of correct, but I agree: it doesn't parse very well. Will fix.

File: src/lib/kimyi.rs [hunk 3]
```diff
@@ -0,0 +1,737 @@
...
+                                                      .unwrap()
+                                                      .clone()));
+        rpr_pstack = p;
+    }
+    in_pstack.reverse();
+
+    (la_idx, vec![smpl_rprs])
+}
+
+fn r3is<TokId: Clone + Copy + Debug + TryFrom<usize> + TryInto<usize> + PartialEq>
```
Comment (ptersilie) [inline src/lib/kimyi.rs]: I guess I'll have to read the paper to know what `r3is`, `r3ir`, `r3d`, etc, are. I guess these are fixed functions from the paper so we can't give them sensible names. But maybe their function can be summarised in a small comment?

Comment (ltratt) [reply src/lib/kimyi.rs]: I think people have to read the paper to understand those... I could tell them what page to read though.

File: src/lib/kimyi.rs [hunk 4]
```diff
@@ -0,0 +1,737 @@
...
+        } else {
+            break;
+        }
+    }
+
+    // Remove all inserts of nonterms which have a minimal sentence cost of 0.
+    let mut j = 0;
+    while j < rprs.len() {
+        if let ParseRepair::InsertNonterm{nonterm_idx} = rprs[j] {
+            let msc = sg.min_sentence_cost(nonterm_idx);
```
Comment (ptersilie) [inline src/lib/kimyi.rs]: I'm personally not bothered by this but I'm unsure if this is ok rust syntax, as `msc` is only used once, so there's no real need to assign this variable.

Comment (ltratt) [reply src/lib/kimyi.rs]: Good point.

File: src/lib/parser.rs [hunk 5]
```diff
@@ -246,21 +277,45 @@ impl<'a, TokId: Clone + Copy + Debug + PartialEq + TryFrom<usize> + TryInto<usiz
...
+       (grm: &YaccGrammar, sgraph: &StateGraph, stable: &StateTable,
+        lexemes: &Vec<Lexeme<TokId>>)
+    -> Result<Node<TokId>, (Node<TokId>, Vec<ParseError<TokId>>)>
+{
+    parse_rcvry(RecoveryKind::KimYi, grm, |_| 1, |_| 1, sgraph, stable, lexemes)
+}
+
+/// Parse the lexemes, specifying a particularly type of error recovery, returning either a parse
+/// tree or a vector of `ParseError`s.
+pub fn parse_rcvry
```
Comment (ptersilie) [inline src/lib/parser.rs:296]: Hah, you really couldn't be bothered to type those 3 extra characters? ;-)

Comment (ltratt) [reply src/lib/parser.rs:296]: I'm not that lazy :) But it makes it easier for function callers...

**Comment by ptersilie** [comment]
Rest looks good. Ready for merging once squashed.

**Comment by ltratt** [comment]
OK, I'm going to see if I can undo the damage I did in the last rebase...

File: src/lib/parser.rs (hunk 5 above)
Comment (ptersilie) [reply src/lib/parser.rs:296]: Does it? I found it easier to type `recovery` than `rcvry`.

Comment (ltratt) [reply src/lib/parser.rs:296]: I can change it if you think it would be better.

Comment (ptersilie) [reply src/lib/parser.rs:296]: It's not the end of the world, but I always find it easier to type full words rather than abbreviations that force me to think about how it was spelled. Then again, most editors have autocomplete.

**Comment by ltratt** [comment]
OK, I think I've moved everything to the right place now. [Thank goodness for the "edit" option in git rebase!] If you agree, please feel free to merge.

**Comment by ptersilie** [comment]
Merged!
